### PR TITLE
feat(config): warn if `globalOnly` option found in repo config

### LINKED ISSUE #24560: Raise config validation warning for any use of global options in repo config
### Describe the proposed change(s).

Renovate should log a config warning if any global only options are used in repo config.

We can later elevate this to a config _error_ / abort run.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes

This PR adds the functionality to warn if global options are found inside the RepoConfig. I did this by modifying the `validateConfig` fn, which is used for config validation. 

First, I added a new parameter `isGlobalConfig` to the function which will be used to check whether the config will be validated as global config or not. Aftre this, I added an if check for global options, based on the `globalOnly` value.
If a config option is `global`:
1. and  `isGlobalConfig:true` -> continue (a separate fn is needed to validate global-options, a task for another PR)
2. and  `isGlobalConfig:false` ->then push a warning to the warning's list.

There was a side-effect of this addition.
The `validateConfig` fn is also used by the `config-validator` when validating config files manually. A user can validate either the FileConfig, RepoConfig or a random-config-file(incl. PresetConfigFiles)

**We know whether to allow global options inside Repo and File config. But for Preset configs it should be upto the user to decide. But adding that functionality is out of scope of this PR. Hence, I have currently allowed global options in the PreetConfigFiles to be in-line with current behaviour.**

By _in-line with current behaviour_ I mean that if we run `pnpm config-validator <fileName>` today, we won't show any warnings for global options in the config file.

[Here's ](https://www.tldraw.com/s/v2_c_mZP8RCqec4qNFLEMJ6lPU?viewport=-675%2C-71%2C2573%2C1250&page=page%3ASIRSoLPU_haoUzBMT_nkm)a flowchart depicting config-validator's old and new flow.

## Context
Closes: https://github.com/renovatebot/renovate/issues/24560

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Please test/show on some real repos

File: lib/config/validation.ts [hunk 1]
```diff
@@ -113,6 +124,13 @@ export async function validateConfig(
       }
     });
   }
+  if (!optionGlobals) {
+    optionGlobals = {};
+    options.forEach((option) => {
+      optionGlobals[option.name] = option.globalOnly;
+    });
```
Comment (viceice) [inline lib/config/validation.ts:131]: use a `Set<string>`, it's much faster for lookup.

use for-of loop.

File: lib/config-validator.ts [hunk 2]
```diff
@@ -20,7 +20,8 @@ async function validate(
   desc: string,
   config: RenovateConfig,
   strict: boolean,
-  isPreset = false
+  isPreset = false,
+  isFileConfig = false
```
Comment (viceice) [inline lib/config-validator.ts:24]:
```suggestion
  isGlobalConfig = false
```
more clear what that means. presets are also files configs.

File: lib/config/validation.ts [hunk 3]
```diff
@@ -27,9 +27,19 @@ const options = getOptions();
...
 const managerList = getManagerList();
 
+// allow these global options in repo config
+const allowedGlobalOptions = [
+  'customEnvVariables',
+  'migratePresets',
+  'productLinks',
+  'secrets',
+  'customizeDashboard',
+];
```
Comment (viceice) [inline lib/config/validation.ts:41]: @rarkins why do we have these excludes? (copied from code below) shouldn't we remove the global only flag from option instead if they are allowed in repo config?

**Comment by RahulGautamSingh** [comment]
TestRun : https://github.com/RahulGautamSingh-testing/config-validation/issues/1

File: lib/config/validation.ts (hunk 3 above)
Comment (rarkins) [reply lib/config/validation.ts:41]: I don't understand why.

These are/were handled separately because they're _objects_, but this doesn't mean all of them are allowed in repo config

Comment (RahulGautamSingh) [reply lib/config/validation.ts:41]: We should keep it for now, and I will remove them in a new PR when new validation is added to validate the config options.

File: lib/config/validation.ts [hunk 4]
```diff
@@ -27,9 +27,19 @@ const options = getOptions();
 
 let optionTypes: Record<string, RenovateOptions['type']>;
 let optionParents: Record<string, RenovateOptions['parent']>;
+let optionGlobals: Set<string>;
 
 const managerList = getManagerList();
 
+// allow these global options in repo config
```
Comment (rarkins) [inline lib/config/validation.ts:34]: These shouldn't not be allowed

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config-validator.ts (hunk 2 above)
Comment (viceice) [reply lib/config-validator.ts:24]: why resolved without a comment? 😕

File: lib/config/validation.spec.ts [hunk 5]
```diff
@@ -28,6 +28,19 @@ describe('config/validation', () => {
       expect(warnings).toMatchSnapshot();
     });
 
+    it('catches global options in repo config', async () => {
+      const config = {
+        binarySource: 'something',
+      };
+      const { errors } = await configValidation.validateConfig(config);
+      expect(errors).toHaveLength(1);
+      expect(errors).toMatchSnapshot([
```
Comment (viceice) [inline lib/config/validation.spec.ts:37]:
```suggestion
      expect(errors).toMatchObject([
```

File: lib/config-validator.ts (hunk 2 above)
Comment (RahulGautamSingh) [reply lib/config-validator.ts:24]: My bad...I think the change didn't get saved before staging and commiting

File: lib/config-validator.ts [hunk 6]
```diff
@@ -134,7 +140,7 @@ type PackageJson = {
         const file = process.env.RENOVATE_CONFIG_FILE ?? 'config.js';
         logger.info(`Validating ${file}`);
         try {
-          await validate(file, fileConfig, strict);
+          await validate(file, fileConfig, strict, true);
```
Comment (rarkins) [inline lib/config-validator.ts:143]: This function takes 5 params, you have provided 4. You are setting isPreset=true when you presumably mean isGlobalConfig=true.

This is a lesson to avoid optional params like this! Please refactor so they are mandatory and passed each time

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/validation.ts [hunk 7]
```diff
@@ -697,3 +706,61 @@ function validateRegexManagerFields(
     }
   }
 }
+
+// basic validation for global config options
```
Comment (rarkins) [inline lib/config/validation.ts:710]: Can this refactoring be done separately/first?

Comment (RahulGautamSingh) [reply lib/config/validation.ts:710]: Okay 👍

Comment (RahulGautamSingh) [reply lib/config/validation.ts:710]: https://github.com/renovatebot/renovate/pull/25218

**Comment by RahulGautamSingh** [comment]
Blocked by : https://github.com/renovatebot/renovate/pull/25218

File: lib/config/validation.spec.ts [hunk 8]
```diff
@@ -912,5 +931,124 @@ describe('config/validation', () => {
         },
       ]);
     });
+
+    describe('globalOptions()', () => {
+      it('validates options of string type', async () => {
+        const config = {
+          binarySource: 'docker',
+          gitUrl: false as never,
```
Comment (viceice) [inline lib/config/validation.spec.ts:939]:
```suggestion
          // @ts-expect-error
          gitUrl: false,
```
use ts-expect-error

we should add this to the dev guide

File: lib/config/validation.spec.ts [hunk 9]
```diff
@@ -912,5 +931,124 @@ describe('config/validation', () => {
...
+            message: 'Configuration option `gitUrl` should be a string',
+            topic: 'Configuration Error',
+          },
+        ]);
+      });
+
+      it('validates options of boolean type', async () => {
+        const config = {
+          unicodeEmoji: false,
+          detectGlobalManagerConfig: 'invalid-type' as never,
```
Comment (viceice) [inline lib/config/validation.spec.ts:959]: same

File: lib/config/validation.spec.ts [hunk 10]
```diff
@@ -912,5 +931,124 @@ describe('config/validation', () => {
...
+            )} (string)`,
+            topic: 'Configuration Error',
+          },
+        ]);
+      });
+
+      it('validates options of integer type', async () => {
+        const config = {
+          prCommitsPerRunLimit: 2,
+          gitTimeout: 'invalid-type' as never,
```
Comment (viceice) [inline lib/config/validation.spec.ts:981]: same

File: lib/config/validation.spec.ts [hunk 11]
```diff
@@ -912,5 +931,124 @@ describe('config/validation', () => {
...
+            )} (string)`,
+            topic: 'Configuration Error',
+          },
+        ]);
+      });
+
+      it('validates options of array type', async () => {
+        const config = {
+          allowedPostUpgradeCommands: ['cmd'],
+          checkedBranches: 'invalid-type' as never,
```
Comment (viceice) [inline lib/config/validation.spec.ts:1003]: same

**Comment by RahulGautamSingh** [comment]
Test Run: TestRun : https://github.com/RahulGautamSingh-testing/config-validation/issues/1

File: lib/config-validator.ts [hunk 12]
```diff
@@ -20,6 +20,7 @@ async function validate(
   desc: string,
   config: RenovateConfig,
   strict: boolean,
+  isGlobalConfig: boolean,
```
Comment (rarkins) [inline lib/config-validator.ts:23]:
```suggestion
  isGlobalConfig: boolean,
  desc: string,
  config: RenovateConfig,
  strict: boolean,
```

Please reorder like this, or refactor the code in a seprate PR so that the isPreset boolean is no longer optional. It's too easy to miss mistakes this way

File: lib/config/validation.ts [hunk 13]
```diff
@@ -97,7 +98,8 @@ export function getParentName(parentPath: string | undefined): string {
 export async function validateConfig(
   config: RenovateConfig,
   isPreset?: boolean,
-  parentPath?: string
+  parentPath?: string,
+  isGlobalConfig?: boolean
```
Comment (rarkins) [inline lib/config/validation.ts:102]:
```suggestion
  isGlobalConfig: boolean,
  parentPath?: string
```

Make it mandatory

File: lib/config/validation.ts [hunk 14]
```diff
@@ -96,6 +97,7 @@ export function getParentName(parentPath: string | undefined): string {
 
 export async function validateConfig(
   config: RenovateConfig,
+  isGlobalConfig: boolean,
```
Comment (rarkins) [inline lib/config/validation.ts:100]:
```suggestion
  isGlobalConfig: boolean,
  config: RenovateConfig,
```

Also change this order to better catch type errors

File: lib/config-validator.ts [hunk 15]
```diff
@@ -112,12 +113,19 @@ type PackageJson = {
...
+          pkgJson.renovate,
+          strict,
+          false
+        );
       }
       if (pkgJson['renovate-config']) {
         logger.info(`Validating package.json > renovate-config`);
         for (const presetConfig of Object.values(pkgJson['renovate-config'])) {
           await validate(
+            true,
```
Comment (rarkins) [inline lib/config-validator.ts:128]: Are you sure this is global config? It's a preset isn't it?

File: lib/config-validator.ts
Comment (rarkins) [inline lib/config-validator.ts:1]: I think we may need to add a flag to the validator to let the user force global or non-global mode in case we get it wrong. For example it's possible that a preset could be used for repo config, or global.

We could autodetect by default, but have an option for them to override

**Review by rarkins — CHANGES_REQUESTED** [review]
Will this change cause a repo to _stop_ due to config error? Or add a warning? I think we should start with warnings in case we get it wrong.

**Comment by RahulGautamSingh** [comment]
> Will this change cause a repo to _stop_ due to config error? Or add a warning? I think we should start with warnings in case we get it wrong.

Warn

**Comment by rarkins** [comment]
Can you show an example of a repo with 2+ warnings and a dependency dashboard?

File: lib/config-validator.ts (hunk 15 above)
Comment (RahulGautamSingh) [reply lib/config-validator.ts:128]: Sorry, it is a mistake.

File: lib/config-validator.ts
Comment (RahulGautamSingh) [reply lib/config-validator.ts:1]: Autodetect will be based on the filename, right? If it's one of the allowed config file names, it's repo config else the global config.

File: lib/config-validator.ts (hunk 15 above)
Comment (RahulGautamSingh) [reply lib/config-validator.ts]: We need a test file for the config validator file, it will catch these mistakes.

**Comment by RahulGautamSingh** [comment]
> Can you show an example of a repo with 2+ warnings and a dependency dashboard?

The warnings are logged only, should I add them to the dependency dashboard?

File: lib/config-validator.ts
Comment (rarkins) [reply lib/config-validator.ts:1]: Yes, but it won't always be perfect

**Comment by rarkins** [comment]
Single line is ok for now

File: lib/config/validation.ts [hunk 16]
```diff
@@ -132,6 +143,18 @@ export async function validateConfig(
         message: `The "${key}" object can only be configured at the top level of a config but was found inside "${parentPath}"`,
       });
     }
+    if (optionGlobals.has(key)) {
+      if (!isGlobalConfig) {
```
Comment (rarkins) [inline lib/config/validation.ts:147]:
```suggestion
    if (optionGlobals.has(key)) {
      if (!isGlobalConfig) {
```

Swap this logic to avoid the .has() check when validating global config

File: lib/config/validation.ts [hunk 17]
```diff
@@ -603,28 +627,13 @@ export async function validateConfig(
...
-                'secrets',
-                'customizeDashboard',
-              ].includes(key)
-            ) {
-              const res = validatePlainObject(val);
-              if (res !== true) {
-                errors.push({
-                  topic: 'Configuration Error',
-                  message: `Invalid \`${currentPath}.${key}.${res}\` configuration: value is not a string`,
-                });
```
Comment (rarkins) [inline lib/config/validation.ts:620]: Where/how is this validation done now? Can this be refactored in a separate PR as it seems unrelated to global config enforcement?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/config/validation.ts (hunk 17 above)
Comment (RahulGautamSingh) [reply lib/config/validation.ts:620]: This check for global options becomes obsolete cause of the above if check for global options, so I removed it.

I have separated this refactor in this PR: https://github.com/renovatebot/renovate/pull/25218

Comment (RahulGautamSingh) [reply lib/config/validation.ts]: There is no need to remove them in this PR. I will remove them in another PR.

**Review by viceice — DISMISSED** [review]
you should update the PR title to match the warning behavior

**Review by rarkins — CHANGES_REQUESTED** [review]
New behavior of the CLI config validator should be described in docs somehow (I forget how much we have right now). For example how Renovate decides whether to evaluate it as a repo or global config if it's a preset.

**Comment by RahulGautamSingh** [comment]
> New behavior of the CLI config validator should be described in docs somehow (I forget how much we have right now). For example how Renovate decides whether to evaluate it as a repo or global config if it's a preset.

Regarding the example you added:
I think its the other way around. Presets are collection of config options only, so whether they are evaluated as global config or repo config will depend on the file they are present in.

**Comment by rarkins** [comment]
Isn't it possible for people to use the validator to point config file at a preset file and validate it directly?

**Comment by RahulGautamSingh** [comment]
It is possible, for a user to do that. They are validated as repo config and this PR does not change this behaviour.

We will need to add an extra flag for it, so that user can decide. Since it is not possible for us to tell from the preset-file name if the preset is used in repo or global config.

I am assuming you mean directly validating a preset file like `default.json` and not the presets which are present in the `extends` array of a repo/global config.

**Comment by rarkins** [comment]
Yes. In this case the behavior is a breaking change and not ready for merge.

**Comment by RahulGautamSingh** [comment]
What should be added in this for it to be merged? Should I a new flag which will allow users to enable/disable the validation of global options present in the file, in a separate PR?

**Comment by rarkins** [comment]
This can't be merged unless any breaking changes are opt-in or removed

File: lib/config-validator.ts [hunk 18]
```diff
@@ -75,7 +76,7 @@ type PackageJson = {
         const parsedContent = await getParsedContent(file);
         try {
           logger.info(`Validating ${file}`);
-          await validate(file, parsedContent, strict);
+          await validate(false, file, parsedContent, strict);
```
Comment (RahulGautamSingh) [inline lib/config-validator.ts:79]:
```suggestion
          await validate(true, file, parsedContent, strict);
```
We can fix the breaking change by allowing global options in the config, if the user is validating a custom file by default.

In another PR we can add a flag to control this behaviour.

**Comment by RahulGautamSingh** [comment]
Test Run: https://github.com/RahulGautamSingh/global-config-validation
I have added the debug logs in the readme section, since the warnings aren't shown in dependency dhasboard.

**Review by viceice — DISMISSED** [review]

File: lib/config/validation.ts [hunk 19]
```diff
@@ -135,6 +146,17 @@ export async function validateConfig(
         message: `The "${key}" object can only be configured at the top level of a config but was found inside "${parentPath}"`,
       });
     }
+    if (!isGlobalConfig) {
+      if (optionGlobals.has(key)) {
+        // token is a global config option and a field of repo config option hostRules.encrypted
+        if (!(key === 'token' && parentPath?.includes('hostRules'))) {
```
Comment (rarkins) [inline lib/config/validation.ts:152]: What about username and password - are they different?

File: lib/config/validation.ts [hunk 20]
```diff
@@ -135,6 +146,17 @@ export async function validateConfig(
         message: `The "${key}" object can only be configured at the top level of a config but was found inside "${parentPath}"`,
       });
     }
+    if (!isGlobalConfig) {
```
Comment (rarkins) [inline lib/config/validation.ts:149]: Couldn't the optionsGlobals be lazily loaded inside here?

File: lib/config/validation.ts (hunk 19 above)
Comment (RahulGautamSingh) [reply lib/config/validation.ts:152]: Missed them, didn't know they could be configured without a parent.

File: lib/config/validation.ts (hunk 20 above)
Comment (RahulGautamSingh) [reply lib/config/validation.ts:149]: Yep it can be.

File: lib/config/validation.ts (hunk 19 above)
Comment (RahulGautamSingh) [reply lib/config/validation.ts]: I looked for all other possible options as well. Here's the result:

By "other possible options" I mean the repo config options which can have fields whose  whose children can have fields same as another `globalOnly` option.

Except for `secrets`, `encrypted`, `hostRules` no other config option will have such fields. Other options either are either validated separately or they have strict restrictions as to which fields are allowed in them.

Among these three options:
- `hostRules` is already handled in this PR.

- `secrets` is validate separately ie. its children/fields won't be recursively passed inside validateConfig` for re-validation. So we don't need extra check for its options.

- Fields inside `encrypted` needs to be handled based on its parent.
If we handle the rest of the options well, `encrypted` doesn't needs to be handled separately.

File: lib/config/validation.ts [hunk 21]
```diff
@@ -135,6 +138,23 @@ export async function validateConfig(
...
+            optionGlobals.add(option.name);
+          }
+        }
+      }
+
+      if (optionGlobals.has(key) && !isFalseGlobal(key, parentPath)) {
+        warnings.push({
+          topic: 'Configuration Error',
+          message: `The "${key}" option is a global option reserved only for bot's global configuration and cannot be configured within repository config file`,
+        });
```
Comment (RahulGautamSingh) [inline lib/config/validation.ts:155]: add `continue` after this, so we move to the next property in the loop, since it's decided that this is invalid global option.

**Review by viceice — APPROVED** [review]

**Comment by RahulGautamSingh** [comment]
I did a big test which validates all the repo config options, in this [repository](https://github.com/RahulGautamSingh-testing/all-renovate-config-options). Since users can only pick parts of what I tested, I'm super confident that whatever they choose will work smoothly in our system!

**Review by rarkins — APPROVED** [review]
